src/main/java/edu/cornell/kfs/pdp/batch/service/impl/PayeeACHAccountExtractServiceImpl.java

```diff
@@ -0,0 +1,691 @@
+        Pattern placeholderPattern = Pattern.compile("\\[([^\\]]+)\\]");
+        Matcher emailMatcher = placeholderPattern.matcher(emailBody.replace("\\n", "\n"));
+        // Use a StringBuffer here, due to the Matcher class not supporting StringBuilder for appending operations.
+        StringBuffer resolvedEmailBody = new StringBuffer(emailBody.length());
+        
+        // Replace all placeholders one by one. The pattern has a single group in it to help with retrieving just the property name and not the brackets.
+        while (emailMatcher.find()) {
+            String propertyName = emailMatcher.group(1);
+            AttributeDefinition attDefinition = dataDictionaryService.getAttributeDefinition(PayeeACHAccount.class.getName(), propertyName);
+            
+            String replacement;
+            // Make sure property exists in data dictionary and is either not potentially sensitive or is flagged as safe to use.
+            if (attDefinition != null) {
+                AttributeSecurity attSecurity = attDefinition.getAttributeSecurity();
+                if (attSecurity != null && (attSecurity.isHide() || attSecurity.isMask() || attSecurity.isPartialMask())
+                        && !allowedAchSecureFields.contains(propertyName)) {
+                    // Replace potentially-sensitive placeholders with an empty string.
+                    replacement = KFSConstants.EMPTY_STRING;
+                } else {
+                    // Replace the placeholder with the property value, or with an empty string if null or invalid.
+                    try {
+                        Object propertyValue = ObjectPropertyUtils.getPropertyValue(achAccount, propertyName);
+                        replacement = ObjectUtils.isNotNull(propertyValue) ? propertyValue.toString() : KFSConstants.EMPTY_STRING;
+                        // If a values finder is defined, use the label from the matching key/value pair instead.
+                        if (attDefinition.getControl() != null && StringUtils.isNotBlank(attDefinition.getControl().getValuesFinderClass())) {
+                            KeyValuesFinder valuesFinder = (KeyValuesFinder) Class.forName(attDefinition.getControl().getValuesFinderClass()).newInstance();
+                            String key = replacement;
+                            replacement = valuesFinder.getKeyLabel(key);
+                            // If the key is in the label, then remove it from the label.
+                            if (attDefinition.getControl().getIncludeKeyInLabel() != null
+                                    && attDefinition.getControl().getIncludeKeyInLabel().booleanValue()) {
+                                // Check for key-and-dash or key-in-parentheses, and remove them if found.
+                                // (Former can come from BO values finders, latter is only for custom values finders that append the keys as such.)

```

Comment:
The code section below these comments has been set up in a manner not using more "if" blocks, to avoid some CheckStyle complaints about the nested "if" block depth being too deep.